Utilities/Files.cs

```diff
@@ -237,4 +277,53 @@ public static bool IsFileLocked(FileInfo file)
             return false;
         }
     }
+
+    // Code from https://www.codeproject.com/Tips/1190802/File-Locking-in-a-Multi-Threaded-Environment
+    // The code maintains a concurrent dictionary of lock objects and keeps track of how many calls are in the queue for
+    // accessing each filename. If the count goes down to 0, the key is removed from the dictionary as a cleanup mechanism.
+    class LockManager
+    {
+        private static readonly ConcurrentDictionary<string, lobj> _locks =
+                new ConcurrentDictionary<string, lobj>();
+
+        public static void GetLock(string filename, Action action)
+        {
+            lock (GetLock(filename))
+            {
+                action();
+                Unlock(filename);
+            }
+        }
+
+        private static lobj GetLock(string filename)
+        {
+            if (_locks.TryGetValue(filename.ToLower(), out var o))
+            {
+                o.count++;
```

Comment:
I am surely missing something here: I assumed that we use locks to prevent concurrent write access to files, but if we increment the refcount and return the lock, we are not blocking the second access.

Reply:
Why would incrementing the ref count unblock anything? The same lock object is being used.

Reply:
Moved to LockManager.cs.

Reply:
You are right, I missed the `lock(` in front of the call to `GetLock`.

Do we then need refcount?
Every lock should be used only by one thread at the time, right?

Reply:
Upon further thinking about this, I see why we need ref counting.
Please, forget my previous comment :)